src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+        return false;
+
+    return true;
+}
+
+struct sol_iio_channel *
+sol_iio_add_channel(struct sol_iio_device *device, const char *name)
+{
+    char path[PATH_MAX + 1];
+    struct sol_iio_channel *channel;
+    struct stat st;
+    int len;
+
+    len = snprintf(path, sizeof(path), CHANNEL_RAW_PATH, device->device_id,
+        name);
+    if (len < 0 || len > PATH_MAX || stat(path, &st)) {
+        SOL_WRN("Could not find channel [%s] for device%d", name,
+            device->device_id);
+        return NULL;
+    }
+
+    channel = calloc(1, sizeof(struct sol_iio_channel));
+    SOL_NULL_CHECK(channel, NULL);
+
+    channel->name = strdup(name);
+    channel->device = device;
+
+    channel_get_scale(channel);
+    channel_get_offset(channel);
+
+    if (device->buffer_enabled) {
+        if (!enable_channel_scan(channel)) {
+            SOL_WRN("Could not enable scanning of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        if (!read_channel_scan_info(channel)) {
+            SOL_WRN("Could not read scanning info of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        if (!read_channel_index_in_buffer(channel)) {
+            SOL_WRN("Could not read index in buffer of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        channel->offset_in_buffer = -1;
+    }
+
+    sol_ptr_vector_append(&channel->device->channels, channel);
+
+    SOL_DBG("channel %s added. scale: %lf - offset: %d - storagebits: %d",
+        channel->name, channel->scale, channel->offset, channel->storagebits);
+
+    return channel;
+
+error:
+    iio_del_channel(channel);
+
+    return NULL;
+}
+
+bool
+sol_iio_read_channel_value(struct sol_iio_channel *channel, double *value)
+{
+    int len, raw_value;
+    char path[PATH_MAX + 1];
+    struct sol_iio_device *device = channel->device;
+
+    if (device->buffer_enabled) {
+        SOL_WRN("device%d is using buffer - use iio_read_buffer_channel_value() instead",
+            device->device_id);
+        return false;
+    }
+
+    len = snprintf(path, sizeof(path), CHANNEL_RAW_PATH, device->device_id,
+        channel->name);
+    if (len < 0) {
+        SOL_WRN("Could not read channel [%s] in device%d", channel->name,
+            device->device_id);
+        return false;
+    }
+
+    len = sol_util_read_file(path, "%d", &raw_value);
+    if (len < 0) {
+        SOL_WRN("Could not read channel [%s] in device%d", channel->name,
+            device->device_id);
+        return false;
+    }
+
+    *value = (raw_value + channel->offset) * channel->scale;
+    return true;
+}
+
+void
+sol_iio_set_reader_cb(struct sol_iio_device *device, sol_iio_reader_cb reader_cb, void *data)
+{
+    device->reader_cb = reader_cb;
+    device->reader_cb_data = data;
+}
```

Comment:
you could create/delete the buffer memory here  instead of each callback of `sol_fd_add()`.

Also, do we want the `sol_fd_add()` even if there is no read callback associated? If not, then you can delete/add the fd monitor here.